packages/react/src/combobox/root/ComboboxRoot.tsx

```diff
@@ -0,0 +1,1135 @@
+      }),
+  ).current;
+
+  const onItemHighlighted = useEventCallback(onItemHighlightedProp);
+  const onOpenChangeComplete = useEventCallback(onOpenChangeCompleteProp);
+
+  const activeIndex = useStore(store, selectors.activeIndex);
+  const selectedIndex = useStore(store, selectors.selectedIndex);
+  const anchorElement = useStore(store, selectors.anchorElement);
+  const positionerElement = useStore(store, selectors.positionerElement);
+  const listElement = useStore(store, selectors.listElement);
+  const triggerElement = useStore(store, selectors.triggerElement);
+  const inputElement = useStore(store, selectors.inputElement);
+  const inline = useStore(store, selectors.inline);
+  const open = inline || openRaw;
+
+  const { mounted, setMounted, transitionStatus } = useTransitionStatus(open);
+  const {
+    openMethod,
+    triggerProps,
+    reset: resetOpenInteractionType,
+  } = useOpenInteractionType(open);
+
+  const initialList = React.useMemo(() => {
+    if (virtualized) {
+      return Array.from({ length: flatFilteredItems.length }, () => null);
+    }
+    return [];
+  }, [virtualized, flatFilteredItems]);
+
+  const listRef = React.useRef<Array<HTMLElement | null>>(initialList);
+  const popupRef = React.useRef<HTMLDivElement | null>(null);
+  const valuesRef = React.useRef<Array<any>>([]);
+  const inputRef = React.useRef<HTMLInputElement | null>(null);
+  const keyboardActiveRef = React.useRef(true);
+  const allowActiveIndexSyncRef = React.useRef(true);
+  const hadInputClearRef = React.useRef(false);
+  const prevQueryRef = React.useRef(query);
+
+  const commitValidation = fieldControlValidation.commitValidation;
+
+  const updateValue = useEventCallback((nextValue: any) => {
+    clearErrors(name);
+    setDirty(nextValue !== validityData.initialValue);
+  });
+
+  const formValue = selectionMode === 'none' ? inputValue : selectedValue;
+
+  useField({
+    id,
+    commitValidation,
+    value: formValue,
+    controlRef: inputRef,
+    name,
+    getValue: () => formValue,
+  });
+
+  useIsoLayoutEffect(() => {
+    if (!virtualized) {
+      return;
+    }
+    // Drop stray nulls
+    listRef.current = initialList.slice();
+    valuesRef.current.length = initialList.length;
+  }, [initialList, virtualized]);
+
+  // Maintain a full value map for virtualized lists so offscreen selections and
+  // navigation use the correct indices even when DOM nodes aren't rendered.
+  useIsoLayoutEffect(() => {
+    if (!virtualized) {
+      return;
+    }
+    valuesRef.current = flatFilteredItems.slice();
+  }, [virtualized, flatFilteredItems]);
+
+  useIsoLayoutEffect(() => {
+    if (
+      prevQueryRef.current === query ||
+      !open ||
+      query === '' ||
+      query === String(defaultInputValue).toLocaleLowerCase()
+    ) {
+      return;
+    }
+
+    setQueryChangedAfterOpen(true);
+  }, [open, query, defaultInputValue]);
+
+  useIsoLayoutEffect(() => {
+    prevQueryRef.current = query;
+  }, [query]);
```

Comment:
Any way to simplify/merge these effects? Just like `useMemo`, calling `useLayoutEffect` is not a free operation.